File: 0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.java
class Solution {
    public int kInversePairs(int n, int k) {
        int MOD = 1000000007;

        int[] dp = new int[k + 1];
        int[] newDp = new int[k + 1];

        dp[0] = 1;

        for (int i = 1; i <= n; i++) {
            int[] prefix = new int[k + 1];
            prefix[0] = dp[0];

            for (int j = 1; j <= k; j++) {
                prefix[j] = (prefix[j - 1] + dp[j]) % MOD;
            }

            for (int j = 0; j <= k; j++) {
                newDp[j] = prefix[j];

                if (j >= i) {
                    newDp[j] = (newDp[j] - prefix[j - i] + MOD) % MOD;
                }
            }

            dp = newDp;
            newDp = new int[k + 1];
        }

        return dp[k];
    }
}

File: 0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        MOD = 10**9 + 7

        dp = [0] * (k + 1)
        dp[0] = 1

        for i in range(1, n + 1):
            prefix = [0] * (k + 1)
            prefix[0] = dp[0]

            for j in range(1, k + 1):
                prefix[j] = (prefix[j - 1] + dp[j]) % MOD

            new_dp = [0] * (k + 1)

            for j in range(k + 1):
                new_dp[j] = prefix[j]

                if j >= i:
                    new_dp[j] = (new_dp[j] - prefix[j - i]) % MOD

            dp = new_dp

        return dp[k]